Better setting example (#386)
TalonCommunity/Wiki: docs/Customization/Talon Framework/settings.md

# Settings

Settings allow you to control some of the parameters of your Python files by changing their value in a .talon file. This can be used to make a Talon user file set easier to customize for end users, such as exposing the background color of a UI element. It can also be useful to have certain settings change when the context changes, by setting them to different values in different .talon files.

> Settings are defined on Modules by calling a Module's `.setting()` method. Each setting has a name, type, default value, and description. A setting can be accessed from Python by calling `settings.get()` with the setting name as the argument. The following example shows how to define a setting in Python and get its contextually dependent value. If you are using Community and Python is the active programming language, you can use the command `snip module setting` to define a setting. Python code accessing settings should only be invoked through a function registered to be called by Talon or an action because accessing a setting through top level Python code has inconsistent behavior.

Author: I tried to improve the wording in response to your suggestions, but I wanted to provide more accurate information. Storing setting values in global variables can make sense in some contexts, such as detecting if a setting changed from the last time an action was called.

> Settings are defined on Modules by calling a Module's `.setting()` method. Each setting has a name, type, default value, and description. A setting's value can be obtained from Python by calling `settings.get()` with the setting name as the argument. The following example shows how to define a setting in Python and get its context-dependent value. If you are using Community and Python is the active programming language, you can use the command `snip module setting` to define a setting. Because settings are undefined during startup, only obtain setting values in a function registered to be called by Talon or an action. Do not use top level (unindented) Python code to run any instructions that obtain setting values.

Reviewer: ```suggestion
Declare a setting on a Module by calling its`.setting()` method. Each setting has a name, type, default value, and description. A setting's value can be obtained from Python by calling `settings.get()` with the setting name as the argument.

The following example shows how to define a setting in Python and get its context-dependent value. If you are using Community and Python is the active programming language, you can use the command `snip module setting` to define a setting. Because settings are undefined during Talon startup, only obtain setting values in a function registered to be called by Talon (e.g., `ready` callback) or an action.
```

From the community backlog session — we really need an example of how to avoid accessing the registry (actions, settings, etc.) from code that is not called in a Talon managed context (actions, appropriate callbacks). We could link to this for multiple places and also document .register.

`setting.py`

```python
from talon import Module, settings, actions

mod = Module()

mod.setting(
    "my_user_file_set_sleep_amount",
    type=int,
    default=200,
    desc="Set the amount of time to sleep in milliseconds",
)

@mod.action_class
class Actions:

>     def my_user_file_set_paste_file_to_next_window():

Reviewer: ```suggestion
    def my_prefix_paste_file_to_next_window():
```

        """Copy the text from the current file to the next window"""
        actions.edit.select_all()
        actions.edit.copy()
        actions.app.window_next()
        # Sleep to avoid  pasting during the window switching process

>         value = settings.get("user.my_user_file_set_sleep_amount")

Reviewer: ```suggestion
        value = settings.get("user.my_prefix_set_sleep_amount")
```

        print(f"The current value of the setting is {value}")
        actions.sleep(f"{value}ms")
        actions.edit.paste()
```

> Note that the name of the setting (the first argument to mod.setting) in the example included the prefix "my_user_file_set". All user defined settings names share the same namespace so it's important to avoid overly generic setting names that may conflict.

Reviewer: ```suggestion
Note that the name of the setting (the first argument to `mod.setting`) in the example included the prefix "my_prefix". All user-defined settings names share the same namespace, so it's important to avoid overly generic setting names that may conflict.
```
